plugins/woocommerce-admin/client/marketplace/utils/functions.tsx

```diff
@@ -30,4 +30,14 @@ const fetchDiscoverPageData = async (): Promise< Array< ProductGroup > > => {
 		} );
 };
 
-export { fetchDiscoverPageData, ProductGroup };
+// Append UTM parameters to a URL, being aware of existing query parameters
+const appendUTMParams = ( url: string, utmParams: object ): string => {
+	const utmString = Object.entries( utmParams )
+		.map( ( [ key, value ] ) => `${ key }=${ value }` )
+		.join( '&' );
+	const separator = url.includes( '?' ) ? '&' : '?';
+
+	return `${ url }${ separator }${ utmString }`;
+};
```

Comment:
Maybe instead of manipulating the URL string, we can use the JavaScript's [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL) and [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) objects. The logic for placing `?` and `&` can be handled by these core APIs. 

`searchParams.set` is [defined](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/set) as "If there are several parameters, the others are deleted." which felt like what we need. Alternatively, there is also an [append](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/append) option. 

Here is what I have in mind:
```suggestion
const appendUTMParams = ( url: string, utmParams: [string, string] ): string => {
	const urlObject = new URL( url );

	if ( ! urlObject ) {
		return url;
	}

	utmParams.forEach( ( [ key, value ] ) => {
		urlObject.searchParams.set( key, value );
	} );

	return urlObject.toString();
};
```

I have also converted `utmParams` from an object to [a tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple). I think `object` is a loose type definition and future devs may find it easier that only a [string, string] is expected. So it can be called with:

```js
vendorUrl = appendUTMParams( vendorUrl, [
	[ 'utm_source', 'extensionsscreen' ],
	[ 'utm_medium', 'product' ],
	[ 'utm_campaign', 'wcaddons' ],
	[ 'utm_content', 'devpartner' ],
] );
```

Reply:
Added with slight mods in https://github.com/woocommerce/woocommerce/pull/39672/commits/28f1396ae8f644489fddfa9ad6a2f300be3d7d7f